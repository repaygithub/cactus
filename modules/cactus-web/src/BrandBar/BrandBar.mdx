---
name: BrandBar
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import BrandBar from './BrandBar'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# BrandBar

The `BrandBar` Components allows yout to render a Logo of your choice and a user menu with a dropdown for actions such as settings and Loggin in/out.

### Try it out

export const code = `<BrandBar
    isProfilePage
    userMenuText'Hershell Jewess'
    logo="https://repay-merchant-resources.s3.amazonaws.com/staging/24bd1970-a677-4ca7-a4d2-e328ddd4691b/repay_logo_new.jpg"
  >
    <BrandBar.UserMenuItem onSelect={() => console.log('Settings'))}>Settings</BrandBar.UserMenuItem>
    <BrandBar.UserMenuItem onSelect={() => console.log('Logout'))}>Logout</BrandBar.UserMenuItem>
  </BrandBar>`

<LiveProvider code={code} scope={{ BrandBar }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Properties

<PropsTable of={BrandBar} />
