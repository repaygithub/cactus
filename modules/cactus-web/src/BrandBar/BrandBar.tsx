import { DescriptiveProfile, NavigationChevronDown } from '@repay/cactus-icons'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

import { ActionBar } from '../ActionBar/ActionBar'
import { OrderHint, useAction } from '../ActionBar/ActionProvider'
import Flex from '../Flex/Flex'
import FocusLock from '../FocusLock/FocusLock'
import { keyDownAsClick } from '../helpers/a11y'
import { PolyFC } from '../helpers/asProps'
import { getTopPosition, getViewport, usePositioning } from '../helpers/positionPopover'
import { classes } from '../helpers/styled'
import {
  border,
  boxShadow,
  insetBorder,
  popupBoxShadow,
  popupShape,
  radius,
  textStyle,
} from '../helpers/theme'
import usePopup, { TogglePopup } from '../helpers/usePopup'
import { useLayout } from '../Layout/Layout'
import { Sidebar } from '../Layout/Sidebar'
import { MenuItemFunc, MenuItemType } from '../MenuItem/MenuItem'
import { SIZES, useScreenSize } from '../ScreenSizeProvider/ScreenSizeProvider'
import { SidebarMenu as ActionMenuPopup } from '../SidebarMenu/SidebarMenu'

interface ItemProps {
  id?: string
  mobileIcon?: React.ReactElement
  orderHint?: OrderHint
  align?: Align
}

interface UserMenuProps {
  id?: string
  isProfilePage?: boolean
  label: React.ReactNode
}

interface ProfileStyleProp {
  $isProfilePage?: boolean
}

interface BrandBarProps extends React.HTMLAttributes<HTMLDivElement> {
  logo?: string | React.ReactElement
}

type Align = 'left' | 'right'

interface DropdownProps extends React.HTMLAttributes<HTMLDivElement> {
  label: React.ReactNode
  listItemSelector?: string
}

interface PopupProps extends React.HTMLAttributes<HTMLDivElement> {
  buttonRef: React.RefObject<HTMLButtonElement>
  popupRef: React.RefObject<HTMLDivElement>
  isOpen: boolean
  $isTiny: boolean
}

// Margin between BrandBar dropdowns and the BrandBar itself.
const DROPDOWN_MARGIN = 8

// Tell Typescript to treat this as a regular functional component,
// even though React knows it's a `forwardRef` component.
const MenuItemFR = React.forwardRef(MenuItemFunc) as any
const MenuItem = MenuItemFR as MenuItemType

export const BrandBarDropdown: React.FC<DropdownProps> = (props) => {
  return <Dropdown {...props} />
}

export const BrandBarUserMenu: React.FC<UserMenuProps> = (props) => {
  const isTiny = SIZES.tiny === useScreenSize()
  const actionButton = isTiny ? <ActionBarUserMenu {...props} /> : null
  const renderButton = useAction(actionButton, 1000, 'cactus-user-menu')
  if (!isTiny) {
    return <UserMenu {...props} />
  }
  return renderButton && <Sidebar layoutRole="brandbar-menu">{renderButton}</Sidebar>
}

export const BrandBarItem: PolyFC<ItemProps, 'div'> = (props) => {
  const { mobileIcon, as: Component, ...rest } = props
  const isTiny = SIZES.tiny === useScreenSize()
  if (isTiny && mobileIcon) {
    return <ActionBar.Panel aria-label="" {...rest} icon={mobileIcon} />
  } else if (Component) {
    return <Component {...(rest as any)} />
  }
  return <>{props.children}</>
}

type BrandBarType = React.FC<BrandBarProps> & {
  Item: typeof BrandBarItem
  UserMenu: typeof BrandBarUserMenu
  UserMenuItem: typeof MenuItem
  Dropdown: typeof BrandBarDropdown
}

export const BrandBar: BrandBarType = ({ logo, children, className, ...props }) => {
  const isTiny = SIZES.tiny === useScreenSize()
  const justify = isTiny ? 'center' : 'flex-end'
  const childrenArray = React.Children.toArray(children)
  const alignedChildren = childrenArray.filter(
    (child) => (child as JSX.Element).props.align !== undefined
  )
  const leftChildren = alignedChildren.filter(
    (child) => (child as JSX.Element).props.align === 'left'
  )
  const rightChildren = alignedChildren.filter(
    (child) => (child as JSX.Element).props.align === 'right'
  )
  const nonAlignedChildren = childrenArray.filter(
    (child) => (child as JSX.Element).props.align === undefined
  )

  const layoutClass = useLayout('brandbar', { grid: 'header' })
  return (
    <StyledBrandBar {...props} className={classes(className, layoutClass)} $isTiny={isTiny}>
      <Flex justifyContent={justify} flexWrap="nowrap">
        <LogoWrapper>{typeof logo === 'string' ? <img alt="Logo" src={logo} /> : logo}</LogoWrapper>
        {!isTiny && leftChildren}
      </Flex>
      {isTiny ? (
        <Flex flexDirection="column" width="100%">
          {(!!leftChildren.length || !!rightChildren.length) && (
            <StyledDropdownRow
              alignItems="flex-end"
              justifyContent="center"
              flexWrap="nowrap"
              width="100%"
            >
              {leftChildren}
              {rightChildren}
            </StyledDropdownRow>
          )}
          <Flex alignItems="flex-end" justifyContent={justify}>
            {nonAlignedChildren}
          </Flex>
        </Flex>
      ) : (
        <Flex justifyContent={justify} flexWrap="nowrap">
          {childrenArray.filter((child) => (child as JSX.Element).props.align !== 'left')}
        </Flex>
      )}
    </StyledBrandBar>
  )
}

BrandBar.displayName = 'BrandBar'
BrandBar.Item = BrandBarItem
BrandBar.UserMenu = BrandBarUserMenu
BrandBar.UserMenuItem = MenuItem
BrandBar.Dropdown = BrandBarDropdown

BrandBar.propTypes = {
  logo: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
}

BrandBarItem.propTypes = {
  mobileIcon: PropTypes.element,
  orderHint: ActionBar.Panel.propTypes?.orderHint,
  align: PropTypes.oneOf(['left', 'right']),
}

BrandBarItem.defaultProps = { orderHint: 'low' }

BrandBarUserMenu.propTypes = {
  label: PropTypes.node.isRequired,
  isProfilePage: PropTypes.bool,
}

BrandBarUserMenu.defaultProps = {
  isProfilePage: false,
}

BrandBarDropdown.propTypes = {
  label: PropTypes.node.isRequired,
}

export default BrandBar

const ActionBarUserMenu: React.FC<UserMenuProps> = ({
  label,
  children,
  isProfilePage,
  id,
  ...rest
}) => {
  const { buttonProps, toggle, popupProps, wrapperProps } = usePopup('menu', {
    id,
    focusControl,
    onWrapperKeyDown: handleArrows,
  })

  const buttonId = buttonProps.id
  popupProps.onClick = React.useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const target = event.target as HTMLDivElement
      if (target.matches('[role="menuitem"], [role="menuitem"] *')) {
        toggle(false, document.getElementById(buttonId as string))
      }
    },
    [toggle, buttonId]
  )

  return (
    <ActionBar.PanelWrapper {...wrapperProps}>
      <ActionMenuButton $isProfilePage={isProfilePage} {...buttonProps} {...rest}>
        <DescriptiveProfile />
      </ActionMenuButton>
      <ActionBar.PanelPopup as={ActionMenuPopup} padding="0" {...popupProps}>
        <PopupHeader>
          <DescriptiveProfile mr="8px" />
          {label}
        </PopupHeader>
        <ActionMenuList>{children}</ActionMenuList>
      </ActionBar.PanelPopup>
    </ActionBar.PanelWrapper>
  )
}

const focusControl = (root: HTMLElement) =>
  Array.from(root.querySelectorAll<HTMLElement>('[role="menuitem"]'))

const handleArrows = (event: React.KeyboardEvent<HTMLElement>, toggle: TogglePopup) => {
  const key = event.key
  if (key === 'ArrowDown') {
    event.preventDefault()
    toggle(true, 1, { shift: true })
  } else if (key === 'ArrowUp') {
    event.preventDefault()
    toggle(true, -1, { shift: true })
  } else if (key === 'End') {
    event.preventDefault()
    toggle(true, -1)
  } else if (key === 'Home') {
    event.preventDefault()
    toggle(true, 0)
  }
}

const positionPopup = (menu: HTMLElement, menuButton: HTMLElement | null) => {
  if (menuButton) {
    const parent = menuButton.offsetParent || getViewport()
    const viewWidth = parent?.getBoundingClientRect().right || window.innerWidth
    const { right, width, bottom, top } = menuButton.getBoundingClientRect()
    const availableBelow = window.innerHeight - bottom
    menu.style.top = `${bottom}px`
    menu.style.right = `${viewWidth - right}px`
    menu.style.minWidth = `${width}px`
    menu.style.maxWidth = '400px'
    menu.style.maxHeight = `${Math.max(availableBelow, top) - DROPDOWN_MARGIN}px`
  }
}

const UserMenu: React.FC<UserMenuProps> = ({ id, label, children, isProfilePage, ...rest }) => {
  const { expanded, buttonProps, toggle, popupProps, wrapperProps } = usePopup('menu', {
    id,
    focusControl,
    onWrapperKeyDown: handleArrows,
  })
  const buttonRef = React.useRef<HTMLButtonElement>(null)
  const listRef = React.useRef<HTMLUListElement>(null)
  usePositioning({
    position: positionPopup,
    visible: expanded,
    ref: listRef,
    anchorRef: buttonRef,
    updateOnScroll: true,
  })
  popupProps.onClick = React.useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const target = event.target as HTMLDivElement
      if (target.matches('[role="menuitem"], [role="menuitem"] *')) {
        toggle(false, buttonRef.current)
      }
    },
    [toggle]
  )
  return (
    <StyledUserMenu {...wrapperProps}>
      <MenuButton {...buttonProps} {...rest} ref={buttonRef}>
        <DescriptiveProfile aria-hidden mr="8px" />
        <span>{label}</span>
        <NavigationChevronDown aria-hidden ml="8px" />
      </MenuButton>
      <MenuList {...popupProps} ref={listRef}>
        {children}
      </MenuList>
    </StyledUserMenu>
  )
}

const StyledUserMenu = styled.div`
  outline: none;
`

const Dropdown: React.FC<DropdownProps> = ({
  label,
  children,
  listItemSelector = '[role="menuitem"], [role="menuitem"] *',
}) => {
  const buttonRef = React.useRef<HTMLButtonElement>(null)
  const portalRef = React.useRef<HTMLDivElement>(null)
  const isTiny = SIZES.tiny === useScreenSize()
  const { expanded, toggle, buttonProps, popupProps, wrapperProps, setFocus } = usePopup('dialog', {
    onWrapperKeyDown: handleArrows,
  })

  const handlePopupKeyDown = (event: React.KeyboardEvent) => {
    const key = event.key
    if (key === 'Enter' || key === ' ') {
      keyDownAsClick(event)
    }
    switch (key) {
      case 'Home':
        setFocus(0)
        break
      case 'End':
        setFocus(-1)
        break
      case 'Escape':
        toggle(false, buttonRef.current)
        break
    }
  }

  const handlePopupClick = React.useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const target = event.target as HTMLDivElement
      if (target.matches(listItemSelector)) {
        toggle(false, buttonRef.current)
      }
    },
    [toggle, listItemSelector]
  )

  return (
    <div {...wrapperProps}>
      <DropdownButton $isTiny={isTiny} ref={buttonRef} {...buttonProps}>
        <span>{label}</span>
        <NavigationChevronDown aria-hidden ml={4} />
      </DropdownButton>
      <DropdownPopup
        buttonRef={buttonRef}
        popupRef={portalRef}
        isOpen={expanded}
        $isTiny={isTiny}
        {...popupProps}
        onKeyDown={handlePopupKeyDown}
        onClick={handlePopupClick}
      >
        {children}
      </DropdownPopup>
    </div>
  )
}

const BasePopup: React.FC<PopupProps> = ({ buttonRef, popupRef, isOpen, ...props }) => {
  const isTiny = SIZES.tiny === useScreenSize()
  const positionPortal = React.useCallback<
    (popover: HTMLElement, target: HTMLElement | null) => void
  >(
    (popover, target) => {
      if (!target || !popover) {
        return
      }
      const targetRect = target.getBoundingClientRect()
      const popoverRect = popover.getBoundingClientRect()
      const buttonWidth = targetRect.width
      const availableBelow = window.innerHeight - targetRect.bottom

      const topPosition = getTopPosition(targetRect, popoverRect)
      popover.style.top = topPosition.top
      popover.style.left = isTiny ? '0px' : `${targetRect.left}px`
      popover.style.minWidth = `${buttonWidth}px`
      popover.style.maxHeight = `${Math.max(availableBelow, targetRect.top) - DROPDOWN_MARGIN}px`
      if (isTiny) {
        popover.style.right = '0px'
      } else {
        popover.style.maxWidth = `${window.innerWidth - targetRect.left}px`
      }
    },
    [isTiny]
  )
  usePositioning({
    anchorRef: buttonRef,
    ref: popupRef,
    visible: isOpen,
    position: positionPortal,
    updateOnScroll: true,
  })
  return <FocusLock ref={popupRef} {...props} />
}

const StyledDropdownRow = styled(Flex)`
  border-top: ${(p) => border(p.theme, 'lightContrast')};
  border-bottom: ${(p) => border(p.theme, 'lightContrast')};
`

const DropdownButton = styled.button<{ $isTiny: boolean }>`
  box-sizing: border-box;
  cursor: pointer;
  font-weight: 600;
  margin: 0;
  height: 100%;
  display: flex;
  align-items: center;
  flex-wrap: nowrap;
  background-color: transparent;
  ${(p) => textStyle(p.theme, 'body')};
  padding: ${(p) => p.theme.space[4]}px;
  border: 0;
  border-left: ${(p) => border(p.theme, 'lightContrast')};
  border-right: ${(p) => border(p.theme, 'lightContrast')};
  outline: none;

  &[aria-expanded='true'] {
    color: ${(p) => p.theme.colors.callToAction};

    ${NavigationChevronDown} {
      transform: scaleY(-1);
    }
  }

  &:focus {
    ${(p) => !p.$isTiny && insetBorder(p.theme, 'callToAction')};
    color: ${(p) => p.theme.colors.callToAction};
  }

  &:hover {
    color: ${(p) => p.theme.colors.callToAction};
  }

  & svg {
    font-size: 12px;
  }
`

const DropdownPopup = styled(BasePopup)`
  display: block;
  &[aria-hidden='true'] {
    display: none;
  }
  box-sizing: border-box;
  position: fixed;
  z-index: 1000;
  overflow-y: auto;
  outline: none;
  margin-top: ${DROPDOWN_MARGIN}px;
  background-color: ${(p): string => p.theme.colors.white};
  ${(p) => popupShape('menu', p.theme.shape)}
  ${(p) => popupBoxShadow(p.theme)}
  outline: none;
  ${(p) => p.$isTiny && 'width: 100%;'}
`

const StyledBrandBar = styled.div<{ $isTiny: boolean }>`
  display: flex;
  ${(p) =>
    p.$isTiny
      ? `
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
  `
      : `
    flex-direction: row;
    justify-content: space-between;
    align-items: stretch;
  `}
  width: 100%;
  ${(p): string => insetBorder(p.theme, 'lightContrast', 'bottom')};
`
const LogoWrapper = styled.div`
  padding: 16px;
  &,
  * {
    display: block;
    max-width: 200px;
    max-height: 60px;
  }
`

const MenuButton = styled.button<ProfileStyleProp>`
  ${(p) => textStyle(p.theme, 'body')};
  height: 100%;
  font-weight: 600;
  margin-right: 5px;
  background-color: transparent;
  border: 0;
  display: flex;
  align-items: center;
  padding: 16px;
  cursor: pointer;
  margin: 0;
  border-radius: 0;
  ${(p) =>
    p.$isProfilePage &&
    `
    color: ${p.theme.colors.callToAction};
    ${insetBorder(p.theme, 'callToAction', 'bottom')};
  `}
  &:focus {
    ${(p) => insetBorder(p.theme, 'callToAction')};
    outline: none;
  }
  &[aria-expanded='true'] {
    ${(p) => insetBorder(p.theme, 'callToAction', 'bottom')};
    color: ${(p): string => p.theme.colors.callToAction};
    ${NavigationChevronDown} {
      transform: scaleY(-1);
    }
  }
  &:hover {
    color: ${(p): string => p.theme.colors.callToAction};
  }
  & svg {
    font-size: 12px;
  }
`

const MenuList = styled.ul`
  display: block;
  box-sizing: border-box;
  position: fixed;
  outline: none;
  z-index: 110;
  overflow-y: auto;
  &[aria-hidden] {
    display: none;
  }
  padding: 8px 0;
  margin-top: ${DROPDOWN_MARGIN}px;
  border-radius: ${radius(8)};
  ${(p) => boxShadow(p.theme, 1) || `border: ${border(p.theme, 'lightContrast')}`};
  background-color: ${(p) => p.theme.colors.white};
  list-style: none;
  [role='menuitem'] {
    box-sizing: border-box;
    position: relative;
    display: block;
    cursor: pointer;
    text-decoration: none;
    word-wrap: break-word;
    overflow-wrap: break-word;
    ${(p) => textStyle(p.theme, 'small')};
    ${(p) => p.theme.colorStyles.standard};
    outline: none;
    padding: 4px 16px;
    text-align: center;
    &:focus {
      background-color: ${(p) => p.theme.colors.lightContrast};
      color: ${(p) => p.theme.colors.callToAction};
    }
    &:hover {
      color: ${(p) => p.theme.colors.callToAction};
    }
  }
`

const ActionMenuList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
  outline: none;
  word-wrap: break-word;
  overflow-wrap: break-word;
  [role='menuitem'] {
    &:hover {
      color: ${(p) => p.theme.colors.callToAction};
    }
  }
`

const ActionMenuButton = styled(ActionBar.Button)<ProfileStyleProp>(
  (p) => p.$isProfilePage && `color: ${p.theme.colors.callToAction};`
)

const PopupHeader = styled.div`
  text-align: right;
  box-sizing: border-box;
  display: block;
  word-wrap: break-word;
  overflow-wrap: break-word;
  padding: 18px 16px;
  ${(p) => textStyle(p.theme, 'body')};
  font-weight: 600;
  color: ${(p) => p.theme.colors.callToAction};
  ${(p) => insetBorder(p.theme, 'callToAction', 'bottom')};
  svg {
    font-size: 12px;
  }
`
