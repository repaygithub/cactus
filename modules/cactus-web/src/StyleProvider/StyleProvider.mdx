---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
