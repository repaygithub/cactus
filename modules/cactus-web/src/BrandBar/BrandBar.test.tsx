import { fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import animationRender from '../../tests/helpers/animationRender'
import { StyleProvider } from '../StyleProvider/StyleProvider'
import BrandBar from './BrandBar'

describe('component: BrandBar', () => {
  describe('mouse usage', (): void => {
    test('can select an action', async (): Promise<void> => {
      const actionOne = jest.fn()
      const { getByText, getByRole } = render(
        <StyleProvider>
          <BrandBar>
            <BrandBar.UserMenu label="Test name">
              <BrandBar.UserMenuItem onClick={actionOne}>Settings</BrandBar.UserMenuItem>
              <BrandBar.UserMenuItem onClick={(): void => console.log('Logout')}>
                Logout
              </BrandBar.UserMenuItem>
            </BrandBar.UserMenu>
          </BrandBar>
          <div>Click me</div>
        </StyleProvider>
      )

      const trigger = getByRole('button', { name: 'Test name' })

      userEvent.click(trigger)
      const settings = getByText('Settings')
      expect(settings).toBeVisible()

      userEvent.click(getByText('Click me'))

      await animationRender()
      expect(settings).not.toBeVisible()

      userEvent.click(trigger)
      userEvent.click(settings)
      await animationRender()

      expect(actionOne).toHaveBeenCalled()
      expect(getByText('Logout')).not.toBeVisible()
      expect(trigger).toHaveFocus()
    })

    test('can interact with a dropdown', async () => {
      const firstOptionClick = jest.fn()
      const { getByText, getByRole } = render(
        <StyleProvider>
          <BrandBar>
            <BrandBar.Item as={BrandBar.Dropdown} label="Test Dropdown">
              <ul>
                <li role="menuitem" onClick={firstOptionClick}>
                  Option 1
                </li>
                <li>Option 2</li>
                <li>Option 3</li>
              </ul>
            </BrandBar.Item>
          </BrandBar>
          <div>Click me</div>
        </StyleProvider>
      )

      const trigger = getByRole('button', { name: 'Test Dropdown' })

      userEvent.click(trigger)
      await animationRender()
      const option1 = getByText('Option 1')

      expect(option1).toBeVisible()

      userEvent.click(getByText('Click me'))
      await animationRender()
      expect(option1).not.toBeVisible()

      userEvent.click(trigger)
      userEvent.click(option1)
      await animationRender()

      expect(firstOptionClick).toHaveBeenCalled()
      expect(getByText('Option 2')).not.toBeVisible()
      expect(trigger).toHaveFocus()
    })

    test('item supports custom item selectors', async () => {
      const { getByText, getByRole } = render(
        <StyleProvider>
          <BrandBar>
            <BrandBar.Item
              as={BrandBar.Dropdown}
              listItemSelector='[role="custom-selector"], [role="custom-selector"] *'
              label="Test Dropdown"
            >
              <ul>
                <li role="menuitem">Option 1</li>
                <li>Option 2</li>
                <li role="custom-selector">Option 3</li>
              </ul>
            </BrandBar.Item>
          </BrandBar>
        </StyleProvider>
      )

      const trigger = getByRole('button', { name: 'Test Dropdown' })

      userEvent.click(trigger)
      const option1 = getByText('Option 1')
      const option2 = getByText('Option 2')
      const option3 = getByText('Option 3')
      userEvent.click(option1)

      expect(option1).toBeVisible()

      userEvent.click(option2)

      expect(option2).toBeVisible()

      userEvent.click(option3)

      expect(option3).not.toBeVisible()
    })
  })

  describe('keyboard usage', (): void => {
    test('can select an action', async (): Promise<void> => {
      const actionOne = jest.fn()
      const actionTwo = jest.fn()

      const { getByText, getByRole } = render(
        <StyleProvider>
          <BrandBar>
            <BrandBar.UserMenu label="Test name">
              <BrandBar.UserMenuItem onClick={actionOne}>Settings</BrandBar.UserMenuItem>
              <BrandBar.UserMenuItem onClick={actionTwo}>Logout</BrandBar.UserMenuItem>
            </BrandBar.UserMenu>
          </BrandBar>
        </StyleProvider>
      )

      const trigger = getByRole('button', { name: 'Test name' })

      fireEvent.keyDown(trigger, { key: 'Enter' })

      const settings = getByText('Settings')
      const logout = getByText('Logout')

      expect(settings).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'End' })

      expect(logout).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Home' })

      expect(settings).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Escape' })

      expect(settings).not.toBeVisible()
      expect(trigger).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Enter' })

      fireEvent.keyDown(document.activeElement as Element, { key: 'ArrowDown' })

      expect(logout).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'ArrowUp' })
      fireEvent.keyDown(document.activeElement as Element, { key: 'Enter' })

      expect(actionOne).toHaveBeenCalled()
      expect(actionTwo).not.toHaveBeenCalled()
      expect(getByText('Logout')).not.toBeVisible()
      expect(trigger).toHaveFocus()
    })

    test('can interact with a dropdown', () => {
      const thirdOptionClick = jest.fn()
      const { getByText, getByRole } = render(
        <StyleProvider>
          <BrandBar>
            <BrandBar.Item as={BrandBar.Dropdown} label="Test Dropdown">
              <ul>
                <li tabIndex={0}>Option 1</li>
                <li tabIndex={0}>Option 2</li>
                <li role="menuitem" onClick={thirdOptionClick} tabIndex={0}>
                  Option 3
                </li>
              </ul>
            </BrandBar.Item>
          </BrandBar>
        </StyleProvider>
      )

      const trigger = getByRole('button', { name: 'Test Dropdown' })
      fireEvent.keyDown(trigger, { key: 'Enter' })
      const option1 = getByText('Option 1')
      const option2 = getByText('Option 2')
      const option3 = getByText('Option 3')

      expect(option1).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'End' })

      expect(option3).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Home' })

      expect(option1).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Escape' })

      expect(option2).not.toBeVisible()
      expect(trigger).toHaveFocus()

      fireEvent.keyDown(document.activeElement as Element, { key: 'Enter' })
      fireEvent.keyDown(document.activeElement as Element, { key: 'ArrowDown' })
      fireEvent.keyDown(document.activeElement as Element, { key: 'ArrowDown' })
      fireEvent.keyDown(document.activeElement as Element, { key: 'Enter' })

      expect(thirdOptionClick).toHaveBeenCalled()
      expect(option2).not.toBeVisible()
      expect(trigger).toHaveFocus()
    })
  })
})
