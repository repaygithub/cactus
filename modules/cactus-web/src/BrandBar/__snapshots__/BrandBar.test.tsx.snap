// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: BrandBar snapshot 1`] = `
.c3 {
  margin-right: 8px;
  vertical-align: middle;
}

.c5 {
  margin-left: 8px;
  vertical-align: middle;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  width: 100%;
  box-shadow: inset 0px -1px 0px 0px hsl(200,29%,90%);
}

.c1 {
  padding: 16px;
  max-width: 200px;
  max-height: 80px;
}

.c1 > * {
  max-width: 100%;
  max-height: 100%;
}

.c2 {
  display: none;
}

@media screen and (min-width:768px) {
  .c0 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media screen and (min-width:768px) {
  .c2 {
    margin-right: 5px;
    background-color: transparent;
    border: 0;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    padding: 16px;
    cursor: pointer;
  }

  .c2:focus {
    box-shadow: inset 0px 0px 0px 1px hsl(200,96%,35%);
    outline: none;
  }

  .c2[aria-expanded='true'] {
    box-shadow: inset 0px -1px 0px 0px hsl(200,96%,35%);
    color: hsl(200,96%,35%);
  }

  .c2[aria-expanded='true'] .c4 {
    -webkit-transform: rotate3d(1,0,0,180deg);
    -ms-transform: rotate3d(1,0,0,180deg);
    transform: rotate3d(1,0,0,180deg);
  }

  .c2:hover svg,
  .c2:hover span {
    color: hsl(200,96%,35%);
  }

  .c2 svg {
    font-size: 12px;
  }

  .c2 span {
    font-weight: 600;
    font-size: 18px;
  }
}

@media screen and (min-width:1024px) {

}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <button
      aria-controls="menu--1"
      aria-haspopup="true"
      class="c2"
      data-reach-menu-button=""
      id="menu-button--menu--1"
      type="button"
    >
      <svg
        class="c3"
        fill="currentcolor"
        height="1em"
        mr="8px"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12,14 C8.66390125,14 6,11.0697086 6,7.5 C6,3.93029138 8.66390125,1 12,1 C15.3360987,1 18,3.93029138 18,7.5 C18,11.0697086 15.3360987,14 12,14 Z M12,12 C14.1867487,12 16,10.0054236 16,7.5 C16,4.99457638 14.1867487,3 12,3 C9.81325125,3 8,4.99457638 8,7.5 C8,10.0054236 9.81325125,12 12,12 Z M2.45149933,22.6626662 C2.13244151,23.1134661 1.50834851,23.2202642 1.05754861,22.9012064 C0.606748712,22.5821485 0.499950658,21.9580555 0.819008481,21.5072556 C3.90518485,17.1467671 7.65108363,14.9179688 12,14.9179688 C16.3564684,14.9179688 20.0355137,17.1550465 22.9722241,21.5273973 C23.2801579,21.9858678 23.1581239,22.6071613 22.6996535,22.9150951 C22.241183,23.223029 21.6198895,23.100995 21.3119556,22.6425246 C18.7258796,18.7922191 15.6435316,16.9179688 12,16.9179688 C8.34891637,16.9179688 5.18498442,18.8004986 2.45149933,22.6626662 Z"
        />
      </svg>
      <span>
        Test name
      </span>
      <svg
        aria-hidden="true"
        class="c4 c5"
        fill="currentcolor"
        height="1em"
        ml="8px"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
          transform="rotate(-90 12 12)"
        />
      </svg>
    </button>
  </div>
</div>
`;
